### PR TITLE
fix(datadog_logs sink): reconstruct `ddtags` if not already in format expected by DD logs intake

### PR DESCRIPTION
**Pull request by neuronull**
Related to https://github.com/vectordotdev/vector/pull/20184 , this change adds missing behavior that is needed as a result of potentially parsing the `ddtags` in the `datadog_agent` source.

### CODE REVIEW

File: src/sinks/datadog/logs/sink.rs [hunk 1]
```diff
@@ -102,6 +103,47 @@ fn normalize_event(event: &mut Event) {
...
+        }
+
+        // now, the tags attribute must be at the event root so we will move it there if
+        // needed and move any conflicting field if any.
+        if !(ddtags_path.prefix == PathPrefix::Event && ddtags_path.path.to_string() == "ddtags") {
+            let desired_path = event_path!("ddtags");
+            // if an existing attribute exists here already, move it so to not overwrite it.
+            // yes, technically the rename path could exist, but technically that could always be the case.
+            if log.contains(desired_path) {
+                let rename_attr = "_RESERVED_tags";
```
Comment (jszwedko) [inline src/sinks/datadog/logs/sink.rs:133]:
```suggestion
                let rename_attr = "_RESERVED_ddtags";
```

?

Comment (neuronull) [reply src/sinks/datadog/logs/sink.rs]: good catch 👍

File: src/sinks/datadog/logs/sink.rs [hunk 2]
```diff
@@ -102,6 +103,47 @@ fn normalize_event(event: &mut Event) {
...
+                    if let Some(bytes) = tag_kv.as_bytes() {
+                        if idx > 0 {
+                            all_tags.push(',');
+                        }
+                        let tag_kv_str = String::from_utf8_lossy(bytes);
+                        all_tags.push_str(&tag_kv_str);
+                    }
+                });
+                log.insert(&ddtags_path, all_tags);
+            }
```
Comment (pront) [inline src/sinks/datadog/logs/sink.rs:123]: Optional suggestion:

```suggestion
        if !tags_arr.is_empty() {
            let all_tags: String = tags_arr.iter().filter_map(|tag_kv| {
                if let Some(bytes) = tag_kv.as_bytes() {
                    Some(String::from_utf8_lossy(bytes).to_string())
                } else {
                    None // Skip invalid UTF-8 bytes
                }
            }).collect::<Vec<String>>().join(",");
            log.insert(&ddtags_path, all_tags);
        }
```

I didn't test the above but it looks slightly cleaner.

**Review by pront — APPROVED** [review]

File: src/sinks/datadog/logs/sink.rs (hunk 2 above)
Comment (neuronull) [reply src/sinks/datadog/logs/sink.rs:123]: Nice, yeah I like that better. More clear.

File: src/sinks/datadog/logs/sink.rs [hunk 3]
```diff
@@ -102,6 +104,47 @@ fn normalize_event(event: &mut Event) {
...
+                    .collect::<Vec<String>>()
+                    .join(",");
+
+                log.insert(&ddtags_path, all_tags);
+            }
+        }
+
+        // now, the tags attribute must be at the event root so we will move it there if
+        // needed and move any conflicting field if any.
+        if !(ddtags_path.prefix == PathPrefix::Event && ddtags_path.path.to_string() == "ddtags") {
```
[the remark this answers is not in the document]
Comment (neuronull) [reply src/sinks/datadog/logs/sink.rs:129]: Unfortunately `event_path!()` does not work.

I think I may have found a better alternative though, while working on the splunk HEC sink.

File: src/sinks/datadog/logs/sink.rs [hunk 4]
```diff
@@ -102,6 +104,47 @@ fn normalize_event(event: &mut Event) {
...
+        if let Some(Value::Array(tags_arr)) = log.get(&ddtags_path) {
+            if !tags_arr.is_empty() {
+                let all_tags: String = tags_arr
+                    .iter()
+                    .filter_map(|tag_kv| {
+                        tag_kv
+                            .as_bytes()
+                            .map(|bytes| String::from_utf8_lossy(bytes).to_string())
+                    })
+                    .collect::<Vec<String>>()
```
[the remark this answers is not in the document]
Comment (neuronull) [reply src/sinks/datadog/logs/sink.rs]: nice, that does work 🔥

File: src/sinks/datadog/logs/sink.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (neuronull) [reply src/sinks/datadog/logs/sink.rs]: this at least looks cleaner 
https://github.com/vectordotdev/vector/pull/20198/commits/be1b1b95577936147ed67c73340930e6f7a24ef9

**Review by bruceg — APPROVED** [review]

**Review by jszwedko — APPROVED** [review]

File: src/sources/datadog_agent/mod.rs [hunk 5]
```diff
@@ -246,35 +247,35 @@ impl SourceConfig for DatadogAgentConfig {
                 Some(LegacyKey::InsertIfEmpty(owned_value_path!("status"))),
                 &owned_value_path!("status"),
                 Kind::bytes(),
-                Some("severity"),
+                Some(meaning::STATUS),
```
[the remark this answers is not in the document]
Comment (neuronull) [reply src/sources/datadog_agent/mod.rs]: Note- we decided to part this out into a separate PR since it is unrelated to the fix at hand.

**Pull request merged**